crates/ironrdp-server/src/server.rs

```diff
@@ -0,0 +1,224 @@
+use std::{io::Cursor, net::SocketAddr};
+
+use anyhow::{Error, Result};
+use bytes::BytesMut;
+use ironrdp_pdu::{
+    self,
+    input::{
+        fast_path::{FastPathInput, FastPathInputEvent},
+        InputEventPdu,
+    },
+    mcs, rdp, Action, PduParsing,
+};
+use ironrdp_tokio::{Framed, FramedRead, FramedWrite, TokioFramed};
+use tokio::{net::TcpListener, select};
+use tokio_rustls::TlsAcceptor;
+
+use crate::{
+    acceptor::{self, BeginResult, ServerAcceptor},
+    builder, capabilities,
+    display::{DisplayUpdate, RdpServerDisplay},
+    encoder::UpdateEncoder,
+    handler::RdpServerInputHandler,
+};
+
+pub struct RdpServer<H, D> {
+    opts: RdpServerOptions,
+    handler: H,
+    display: D,
+}
+
+#[derive(Clone)]
+pub struct RdpServerOptions {
+    pub addr: SocketAddr,
+    pub security: RdpServerSecurity,
+}
+
+#[derive(Clone)]
+pub enum RdpServerSecurity {
+    None,
+    SSL(TlsAcceptor),
+}
+
+impl<H, D> RdpServer<H, D>
+where
+    H: RdpServerInputHandler,
+    D: RdpServerDisplay,
+{
+    pub fn new(opts: RdpServerOptions, handler: H, display: D) -> Self {
+        Self { opts, handler, display }
+    }
+
+    pub fn builder() -> builder::RdpServerBuilder<builder::WantsAddr, H, D> {
+        builder::RdpServerBuilder::new()
+    }
+
+    pub async fn run(&mut self) -> Result<(), Error> {
+        let listener = TcpListener::bind(self.opts.addr).await?;
+
+        while let Ok((stream, _)) = listener.accept().await {
+            let size = self.display.size().await;
+            let capabilities = capabilities::capabilities(&self.opts, size.clone());
+            let mut acceptor = ServerAcceptor::new(self.opts.clone(), size, capabilities);
+
+            match acceptor::accept_begin(stream, &mut acceptor).await {
+                Ok(BeginResult::ShouldUpgrade(stream)) => {
+                    let upgraded = acceptor::upgrade(&self.opts.security, stream).await?;
+                    let framed = TokioFramed::new(upgraded);
+                    let (framed, _) = acceptor::accept_finalize(framed, &mut acceptor).await?;
+                    self.client_loop(framed).await?;
+                }
+
+                Ok(BeginResult::Continue(framed)) => {
+                    let (framed, _) = acceptor::accept_finalize(framed, &mut acceptor).await?;
+                    self.client_loop(framed).await?;
+                }
+
+                Err(e) => {
+                    eprintln!("connection error: {:?}", e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn client_loop<S>(&mut self, mut framed: Framed<S>) -> Result<()>
+    where
+        S: FramedWrite + FramedRead,
+    {
+        let mut buffer = vec![0u8; 8192 * 8192];
```

Comment:
Same comment about hardcoded buffer size as above

Reply:
For this buffer, I changed it to start with 4096 bytes and get resized accordingly when it needs to hold a larger amount. Also, the fragments of the display updates are sent one at a time, which should greatly reduce the required size of this buffer.